### ISSUE TITLE
Crashes Safari if attempting to use ui-options object

### ISSUE THREAD

**Issue by xmlilley** [angular-ui/ui-mask#38]
So, there's a really nice undocumented feature, which is that you can override the behavior that causes the input's `$viewValue` to clear when the input is blurred. Simply set `ui-options="{clearOnBlur: false}"` as an attribute where you invoke the directive, and _voilà!_

Except in Safari. 

In Safari, it breaks everything and you get this stack-trace:

```
Error: Requested keys of a value that is not an object.
keys@[native code]
extend@http://localhost:3000/bower_components/angular/angular.js:414:29
http://localhost:3000/bower_components/angular-ui-mask/dist/mask.js:126:67
uiMaskLinkingFunction@http://localhost:3000/bower_components/angular-ui-mask/dist/mask.js:132:39
invokeLinkFn@http://localhost:3000/bower_components/angular/angular.js:8290:15
nodeLinkFn@http://localhost:3000/bower_components/angular/angular.js:7800:23
compositeLinkFn@http://localhost:3000/bower_components/angular/angular.js:7149:23
compositeLinkFn@http://localhost:3000/bower_components/angular/angular.js:7152:24
publicLinkFn@http://localhost:3000/bower_components/angular/angular.js:7028:45
boundTranscludeFn@http://localhost:3000/bower_components/angular/angular.js:7167:28
controllersBoundTransclude@http://localhost:3000/bower_components/angular/angular.js:7827:35
ngIfWatchAction@http://localhost:3000/bower_components/angular/angular.js:22205:26
$digest@http://localhost:3000/bower_components/angular/angular.js:14397:31
$apply@http://localhost:3000/bower_components/angular/angular.js:14660:31
http://localhost:3000/bower_components/angular/angular.js:25880:23
dispatch@http://localhost:3000/bower_components/jquery/dist/jquery.js:4435:14
```

(I have to wonder if the reason this feature isn't documented is because it breaks on Safari.)

This is an angular error that's hard to find documentation on. There are a [couple](https://github.com/angular/angular.js/issues/11427) [mentions](https://github.com/angular/angular.js/issues/10949) of it on Angular's issues, but nobody puts their finger on the cause, I don't think. The cause appears to be: trying to use `angular.extend()` to coerce a non-object into a object. 

In here (approx. lines 108-122):
```js
// we can't use angular.copy nor angular.extend, they lack the power to do a deep merge
linkOptions = (function(original, current) {
  for (var i in original) {
    if (Object.prototype.hasOwnProperty.call(original, i)) {
      if (current[i] === undefined) {
        current[i] = angular.copy(original[i]);
      } else {
        // HERE's THE PROBLEM: We're calling angular.extend() on something that's not an object.
        angular.extend(current[i], original[i]);
      }
    }
  }
  return current;
})(options, linkOptions[0]);

```
we're trying to augment the default options object with any values that were supplied by the user, such as my override for `clearOnBlur`. I'm unclear on why the comment says we can't use `angular.extend()` here. In my testing, it seems to work fine. 

But, if we want to stick with our current formula, I believe we can solve the problem by just dropping a few extraneous lines. Replace the block above with:
```js
linkOptions = (function(original, current) {
    for (var i in original) {
        if (Object.prototype.hasOwnProperty.call(original, i)) {
            if (current[i] === undefined) {
                current[i] = angular.copy(original[i]);
            }
        }
    }
    return current;
})(options, linkOptions[0]);
```

Wherever `current[i]` is not undefined, we simply don't need to bother to replace it, let alone by using angular.extend(). Problem solved on Safari, and everything appears to work as intended.

**Comment by xmlilley** [comment]
Even simpler is if we can just use angular.extend(). In my testing, it _seems_ to work as I expect, but I don't claim to have mastered the whole library. I'm replacing lines   with:

```
if (angular.isObject(linkOptions[0])) {
  linkOptions = angular.extend(options, linkOptions[0]);
}
```

**Comment by PowerKiKi** [comment]
Closed via #39
